pkg/sampling/oteltracestate_test.go

```diff
@@ -0,0 +1,336 @@
+			} else {
+				require.NoError(t, err)
+			}
+			if test.rval != ns {
+				require.True(t, otts.HasRValue())
+				require.Equal(t, test.rval, otts.RValue())
+			} else {
+				require.False(t, otts.HasRValue(), "should have no r-value: %s", otts.RValue())
+			}
+			if test.tval != ns {
+				require.True(t, otts.HasTValue())
+				require.Equal(t, test.tval, otts.TValue())
+			} else {
+				require.False(t, otts.HasTValue(), "should have no t-value: %s", otts.TValue())
+			}
+			var expect []KV
+			for _, ex := range test.extra {
+				k, v, _ := strings.Cut(ex, ":")
+				expect = append(expect, KV{
+					Key:   k,
+					Value: v,
+				})
+			}
+			require.Equal(t, expect, otts.ExtraValues())
+
+			if test.expectErr != nil {
+				return
+			}
+			// on success Serialize() should not modify
+			// test by re-parsing
+			var w strings.Builder
+			require.NoError(t, otts.Serialize(&w))
+			cpy, err := NewOTelTraceState(w.String())
+			require.NoError(t, err)
+			require.Equal(t, otts, cpy)
+		})
+	}
+}
+
+func TestUpdateTValueWithSampling(t *testing.T) {
+	type testCase struct {
+		// The input otel tracestate; no error conditions tested
+		in string
+
+		// The incoming adjusted count; defined whether
+		// t-value is present or not.
+		adjCountIn float64
+
+		// the update probability; threshold and tvalue are
+		// derived from this
+		prob float64
+
+		// when update error is expected
+		updateErr error
+
+		// output t-value
+		out string
+
+		// output adjusted count
+		adjCountOut float64
+	}
+	for _, test := range []testCase{
+		// 8/16 in, sampled at (0x10-0xe)/0x10 = 2/16 => adjCount 8
+		{"th:8", 2, 0x2p-4, nil, "th:e", 8},
+
+		// 8/16 in, sampled at 14/16 => no update, adjCount 2
+		{"th:8", 2, 0xep-4, nil, "th:8", 2},
+
+		// 1/16 in, 50% update (error)
+		{"th:f", 16, 0x8p-4, ErrInconsistentSampling, "th:f", 16},
+
+		// 1/1 sampling in, 1/16 update
+		{"th:0", 1, 0x1p-4, nil, "th:f", 16},
+
+		// no t-value in, 1/16 update
+		{"", 0, 0x1p-4, nil, "th:f", 16},
+
+		// none in, 100% update
+		{"", 0, 1, nil, "th:0", 1},
+
+		// 1/2 in, 100% update (error)
+		{"th:8", 2, 1, ErrInconsistentSampling, "th:8", 2},
+
+		// 1/1 in, 0x1p-56 update
+		{"th:0", 1, 0x1p-56, nil, "th:ffffffffffffff", 0x1p56},
+
+		// 1/1 in, 0x1p-56 update
+		{"th:0", 1, 0x1p-56, nil, "th:ffffffffffffff", 0x1p56},
+
+		// 2/3 in, 1/3 update.  Note that 0x555 + 0xaab = 0x1000.
+		{"th:555", 1 / (1 - 0x555p-12), 0x555p-12, nil, "th:aab", 1 / (1 - 0xaabp-12)},
+	} {
+		t.Run(test.in+"/"+test.out, func(t *testing.T) {
+			otts := OTelTraceState{}
+			if test.in != "" {
+				var err error
+				otts, err = NewOTelTraceState(test.in)
+				require.NoError(t, err)
+			}
+
+			require.Equal(t, test.adjCountIn, otts.AdjustedCount())
+
+			newTh, err := ProbabilityToThreshold(test.prob)
+			require.NoError(t, err)
+
+			upErr := otts.UpdateTValueWithSampling(newTh, newTh.TValue())
+
+			if test.updateErr != nil {
+				require.Equal(t, test.updateErr, upErr)
+			}
```

Comment:
```suggestion
			} else {
				require.NoError(t, upErr)
			}
```

Reply:
Actually, you can just skip the if, and do a:

```
require.Equal(t, test.updateErr, upErr)
```

Reply:
Thanks. Even better -- done.